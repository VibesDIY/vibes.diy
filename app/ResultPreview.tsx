import {
  SandpackCodeEditor,
  SandpackLayout,
  SandpackPreview,
  SandpackProvider,
} from '@codesandbox/sandpack-react';
import { useState } from 'react';

interface ResultPreviewProps {
  code: string;
  dependencies?: Record<string, string>;
  onShare?: () => void;
  shareStatus?: string;
}

const indexHtml = `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.tailwindcss.com"></script>
    <script>
      tailwind.config = {
        darkMode: 'class',
        theme: {
          extend: {
            colors: {
              'light-primary': '#2C2C2C',
              'light-secondary': '#2C2C2C',
              'light-decorative-00': '#EBEAEA',
              'light-decorative-01': '#E0DEDE',
              'light-decorative-02': '#2C2C2C',
              'light-background-00': '#FFFFFF',
              'light-background-01': '#F5F5F5',
              'light-background-02': '#F0F0F0',
              'dark-primary': '#FFFFFF',
              'dark-secondary': '#FFFFFF',
              'dark-decorative-00': '#302F30',
              'dark-decorative-01': '#414141',
              'dark-decorative-02': '#FFFFFF',
              'dark-background-00': '#171616',
              'dark-background-01': '#201F20',
              'dark-background-02': '#201F20',
              'accent-00-light': '#F9A100',
              'accent-01-light': '#F58709',
              'accent-02-light': '#F16C12',
              'accent-03-light': '#EE521C',
              'accent-00-dark': '#FFAA0F',
              'accent-01-dark': '#FF8F0F',
              'accent-02-dark': '#FF7119',
              'accent-03-dark': '#FF612A',
            }
          }
        }
      }
    </script>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.jsx"></script>
  </body>
</html>`;

const defaultCode = `export default function App() {
  return (
    <div className="h-screen flex flex-col items-center justify-center bg-light-background-01 dark:bg-dark-background-01 gap-6">
      <div className="w-32 h-32 relative">
        <svg 
          viewBox="6000 6000 5000 5000"
          className="w-full h-full animate-[float_3s_ease-in-out_infinite]"
          style={{ 
            shapeRendering: 'geometricPrecision',
            textRendering: 'geometricPrecision',
            imageRendering: 'optimizeQuality',
            fillRule: 'evenodd',
            clipRule: 'evenodd',
            filter: 'drop-shadow(0 0 10px rgba(238, 82, 28, 0.2))'
          }}
        >
          <style>
            {
              \`@keyframes float {
                0%, 100% { transform: translateY(0px); }
                50% { transform: translateY(-10px); }
              }
              .fil1 { fill: none; }
              .fil3 { fill: #EE521C; }
              .fil2 { fill: #F16C12; }
              .fil4 { fill: #F58709; }
              .fil5 { fill: #F9A100; }
              .fil0 { fill: white; }\`
            }
          </style>
          <g>
            <g>
              <line className="fil1" x1="8333" y1="6034" x2="6342" y2="9483"/>
              <polygon className="fil2" points="8997,7183 8391,7021 7669,7184 7006,8333 7006,8333 7489,8468 8333,8333"/>
              <path className="fil3" d="M7669 7183l647 0 681 0c0,-491 -267,-920 -663,-1149l-1 0 -664 1149z"/>
              <path className="fil4" d="M8333 8333l-1327 0c0,0 0,0 0,1 0,0 -1,0 -1,0l-663 1149 775 257 552 -257 664 -1149 0 -1zm664 1150l594 230 733 -230 1 0c0,-491 -267,-920 -664,-1150l0 0 -664 1150z"/>
              <path className="fil5" d="M7669 9483l-1327 0 664 1150 0 0 1327 0c-397,-230 -664,-659 -664,-1150l0 0zm2656 0l-1328 0 -664 1150 1328 0 664 -1150z"/>
            </g>
          </g>
        </svg>
      </div>
      <div className="text-center px-4">
        <h1 className="text-3xl font-semibold text-light-primary dark:text-dark-primary">
          Send a message to generate your app.
        </h1>
      </div>
    </div>
  );
}`;

function ResultPreview({ code, dependencies = {}, onShare, shareStatus }: ResultPreviewProps) {
  const [activeView, setActiveView] = useState<'preview' | 'code'>('preview');

  console.log(dependencies);
  return (
    <div className="h-full" style={{ overflow: 'hidden' }}>
      <div className="border-light-decorative-00 dark:border-dark-decorative-00 bg-light-background-00 dark:bg-dark-background-00 flex items-center justify-between border-b p-4">
        <div className="bg-light-decorative-00 dark:bg-dark-decorative-00 flex space-x-1 rounded-lg p-1 shadow-sm">
          <button
            type="button"
            onClick={() => setActiveView('preview')}
            className={`flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors ${
              activeView === 'preview'
                ? 'bg-light-background-00 dark:bg-dark-background-00 text-light-primary dark:text-dark-primary shadow-sm'
                : 'text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01'
            }`}
            aria-label="Switch to preview"
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="h-4 w-4"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <title>Preview icon</title>
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
              />
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M2.458 12C3.732 7.943 7.523 5 12 5c4.478 0 8.268 2.943 9.542 7-1.274 4.057-5.064 7-9.542 7-4.477 0-8.268-2.943-9.542-7z"
              />
            </svg>
            <span>Preview</span>
          </button>
          <button
            type="button"
            onClick={() => setActiveView('code')}
            className={`flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors ${
              activeView === 'code'
                ? 'bg-light-background-00 dark:bg-dark-background-00 text-light-primary dark:text-dark-primary shadow-sm'
                : 'text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01'
            }`}
            aria-label="Switch to code editor"
          >
            <svg
              xmlns="http://www.w3.org/2000/svg"
              className="h-4 w-4"
              fill="none"
              viewBox="0 0 24 24"
              stroke="currentColor"
            >
              <title>Code icon</title>
              <path
                strokeLinecap="round"
                strokeLinejoin="round"
                strokeWidth={2}
                d="M10 20l4-16m4 4l4 4-4 4M6 16l-4-4 4-4"
              />
            </svg>
            <span>Code</span>
          </button>
        </div>

        {onShare && (
          <div className="flex items-center gap-2">
            {shareStatus && (
              <div className="animate-fade-in bg-accent-00-light dark:bg-accent-00-dark text-light-primary dark:text-dark-primary rounded-lg px-3 py-1 text-sm">
                {shareStatus}
              </div>
            )}
            <div className="bg-light-decorative-00 dark:bg-dark-decorative-00 flex space-x-1 rounded-lg p-1 shadow-sm">
              <button
                type="button"
                onClick={onShare}
                className="text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01 flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors"
                aria-label="Share app"
              >
                <svg
                  xmlns="http://www.w3.org/2000/svg"
                  className="h-4 w-4"
                  fill="none"
                  viewBox="0 0 24 24"
                  stroke="currentColor"
                >
                  <title>Share icon</title>
                  <path
                    strokeLinecap="round"
                    strokeLinejoin="round"
                    strokeWidth={2}
                    d="M8.684 13.342C8.886 12.938 9 12.482 9 12c0-.482-.114-.938-.316-1.342m0 2.684a3 3 0 110-2.684m0 2.684l6.632 3.316m-6.632-6l6.632-3.316m0 0a3 3 0 105.367-2.684 3 3 0 00-5.367 2.684zm0 9.316a3 3 0 105.368 2.684 3 3 0 00-5.368-2.684z"
                  />
                </svg>
                <span>Share</span>
              </button>
            </div>
          </div>
        )}
      </div>
      <SandpackProvider
        key={code}
        template="vite-react"
        options={{
          externalResources: ['https://cdn.tailwindcss.com'],
        }}
        customSetup={{
          dependencies: {
            ...dependencies,
            'use-fireproof': '0.20.0-dev-preview-41',
            '@adviser/cement': 'latest',
          },
        }}
        files={{
          '/index.html': {
            code: indexHtml,
            hidden: true,
          },
          '/App.jsx': {
            code: code || defaultCode,
            active: true,
          },
        }}
        theme="light"
      >
        <SandpackLayout className="h-full" style={{ height: 'calc(100vh - 49px)' }}>
          <div
            style={{
              display: activeView === 'preview' ? 'block' : 'none',
              height: '100%',
              width: '100%',
            }}
          >
            <SandpackPreview
              showNavigator={false}
              showOpenInCodeSandbox={false}
              showRefreshButton={true}
              showRestartButton={false}
              showOpenNewtab={false}
              className="h-full w-full"
              style={{ height: '100%' }}
            />
          </div>
          <div
            style={{
              display: activeView === 'code' ? 'block' : 'none',
              height: '100%',
              width: '100%',
            }}
          >
            <SandpackCodeEditor style={{ height: '100%' }} />
          </div>
        </SandpackLayout>
      </SandpackProvider>
    </div>
  );
}

export default ResultPreview;
