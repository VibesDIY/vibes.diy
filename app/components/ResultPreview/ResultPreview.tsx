import {
  SandpackCodeEditor,
  SandpackLayout,
  SandpackPreview,
  SandpackProvider,
  useSandpack,
} from '@codesandbox/sandpack-react';
import { useState, useEffect, useCallback, useRef, useMemo } from 'react';
import { sandpackDependencies } from '../../utils/versions';
import WelcomeScreen from './WelcomeScreen';
import SandpackEventListener from './SandpackEventListener';
import SandpackScrollController from './SandpackScrollController';

interface ResultPreviewProps {
  code: string;
  streamingCode?: string;
  dependencies?: Record<string, string>;
  onShare?: () => void;
  shareStatus?: string;
  isSharedApp?: boolean;
  completedMessage?: string;
  currentMessage?: { content: string };
  currentStreamContent?: string;
  onScreenshotCaptured?: (screenshotData: string) => void;
  initialView?: 'code' | 'preview';
  sessionId?: string;
}

const indexHtml = `<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script src="https://cdn.tailwindcss.com"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/html2canvas/1.4.1/html2canvas.min.js"></script>
    <script>
      tailwind.config = {
        darkMode: 'class',
        theme: {
          extend: {
            colors: {
              'light-primary': '#2C2C2C',
              'light-secondary': '#2C2C2C',
              'light-decorative-00': '#EBEAEA',
              'light-decorative-01': '#E0DEDE',
              'light-decorative-02': '#2C2C2C',
              'light-background-00': '#FFFFFF',
              'light-background-01': '#F5F5F5',
              'light-background-02': '#F0F0F0',
              'dark-primary': '#FFFFFF',
              'dark-secondary': '#FFFFFF',
              'dark-decorative-00': '#302F30',
              'dark-decorative-01': '#414141',
              'dark-decorative-02': '#FFFFFF',
              'dark-background-00': '#171616',
              'dark-background-01': '#201F20',
              'dark-background-02': '#201F20',
              'accent-00-light': '#F9A100',
              'accent-01-light': '#F58709',
              'accent-02-light': '#F16C12',
              'accent-03-light': '#EE521C',
              'accent-00-dark': '#FFAA0F',
              'accent-01-dark': '#FF8F0F',
              'accent-02-dark': '#FF7119',
              'accent-03-dark': '#FF612A',
            }
          }
        }
      }

      function captureScreenshot() {
        html2canvas(document.body).then(canvas => {
          const dataURI = canvas.toDataURL();
          window.parent.postMessage({ screenshot: dataURI }, '*');
        });
      }
      
      // Automatically capture screenshot when page is fully loaded
      window.addEventListener('load', function() {
        // Wait a short moment for any final rendering
        setTimeout(captureScreenshot, 500);
      });
    </script>
  </head>
  <body>
    <div id="root"></div>
    <script type="module" src="/index.jsx"></script>
  </body>
</html>`;

const defaultCode = '';

function ResultPreview({
  code,
  streamingCode = '',
  dependencies = {},
  onShare,
  shareStatus,
  isSharedApp,
  completedMessage,
  currentMessage,
  currentStreamContent,
  onScreenshotCaptured,
  initialView = 'preview',
  sessionId,
}: ResultPreviewProps) {
  const [activeView, setActiveView] = useState<'preview' | 'code'>(initialView);
  const [displayCode, setDisplayCode] = useState(code || defaultCode);
  const [appStartedCount, setAppStartedCount] = useState(0);
  const [bundlingComplete, setBundlingComplete] = useState(true);
  const [showWelcome, setShowWelcome] = useState(true);
  const [isDarkMode, setIsDarkMode] = useState(false);
  const codeEditorRef = useRef<HTMLDivElement>(null);
  const [lockCodeView, setLockCodeView] = useState(false);
  const filesRef = useRef({
    '/index.html': {
      code: indexHtml,
      hidden: true,
    },
    '/App.jsx': {
      code: code || defaultCode,
      active: true,
    },
  });

  // Simplify streaming detection - just check if the streamingCode exists
  const hasStreamingContent = Boolean(streamingCode && streamingCode.length > 0);

  useEffect(() => {
    const prefersDarkMode = window.matchMedia('(prefers-color-scheme: dark)').matches;
    setIsDarkMode(prefersDarkMode);

    const mediaQuery = window.matchMedia('(prefers-color-scheme: dark)');
    const handleChange = (e: MediaQueryListEvent) => {
      setIsDarkMode(e.matches);
    };

    mediaQuery.addEventListener('change', handleChange);
    return () => mediaQuery.removeEventListener('change', handleChange);
  }, []);

  useEffect(() => {
    const handleMessage = (event: MessageEvent) => {
      if (event.data && event.data.screenshot) {
        const screenshotData = event.data.screenshot;

        if (onScreenshotCaptured) {
          onScreenshotCaptured(screenshotData);
        }
      }
    };

    window.addEventListener('message', handleMessage);
    return () => window.removeEventListener('message', handleMessage);
  }, [onScreenshotCaptured]);

  // Simplified code update logic - always use the most up-to-date code
  useEffect(() => {
    // Clean the code and add whitespace
    const processCode = (sourceCode: string) => {
      return cleanCodeBeforeImport(sourceCode) + '\n\n\n\n\n\n\n\n\n\n';
    };

    // IMPORTANT: Prioritize streaming code when it exists, otherwise use static code
    const codeToUse = streamingCode || code;
    
    if (codeToUse) {
      console.log('ResultPreview: Updating code, lengths - streamingCode:', 
                 streamingCode?.length || 0, 'code:', code?.length || 0);
      const processedCode = processCode(codeToUse);
      setDisplayCode(processedCode);
      
      filesRef.current = {
        ...filesRef.current,
        '/App.jsx': {
          code: processedCode,
          active: true,
        },
      };
      
      setShowWelcome(false);
      
      // Show code view during streaming
      if (hasStreamingContent) {
        setActiveView('code');
        setLockCodeView(true);
      } else {
        setLockCodeView(false);
      }
    }
  }, [code, streamingCode]);

  // Create a unique key for SandpackProvider that changes when relevant props change
  const sandpackKey = useMemo(() => {
    // Use the actual content that should trigger a remount
    return `${sessionId || 'default'}-${hasStreamingContent ? 'streaming' : 'static'}-${code.length}`;
  }, [sessionId, hasStreamingContent, code]);

  return (
    <div className="h-full" style={{ overflow: 'hidden' }}>
      <style>
        {`
          @keyframes spin-slow {
            from {
              transform: rotate(0deg);
            }
            to {
              transform: rotate(360deg);
            }
          }
          .animate-spin-slow {
            animation: spin-slow 1s linear infinite;
          }
        `}
      </style>
      <div className="border-light-decorative-00 dark:border-dark-decorative-00 bg-light-background-00 dark:bg-dark-background-00 flex min-h-[4rem] items-center justify-between border-b px-6 py-4">
        {!showWelcome ? (
          <div className="bg-light-decorative-00 dark:bg-dark-decorative-00 flex space-x-1 rounded-lg p-1 shadow-sm">
            <button
              type="button"
              onClick={() => setActiveView('preview')}
              className={`flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors ${
                activeView === 'preview'
                  ? 'bg-light-background-00 dark:bg-dark-background-00 text-light-primary dark:text-dark-primary shadow-sm'
                  : 'text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01'
              }`}
              aria-label="Switch to preview"
            >
              <svg
                xmlns="http://www.w3.org/2000/svg"
                className={`h-4 w-4 ${!bundlingComplete && !hasStreamingContent ? 'animate-spin-slow' : ''}`}
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
              >
                <title>Preview icon</title>
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M15 12a3 3 0 11-6 0 3 3 0 016 0z"
                />
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M2.458 12C3.732 7.943 7.523 5 12 5c4.478 0 8.268 2.943 9.542 7-1.274 4.057-5.064 7-9.542 7-4.477 0-8.268-2.943-9.542-7z"
                />
              </svg>
              <span>Preview</span>
            </button>
            <button
              type="button"
              onClick={() => {
                setActiveView('code');
                setShowWelcome(false);
              }}
              className={`flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors ${
                activeView === 'code'
                  ? 'bg-light-background-00 dark:bg-dark-background-00 text-light-primary dark:text-dark-primary shadow-sm'
                  : 'text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01'
              }`}
              aria-label="Switch to code editor"
            >
              <svg
                xmlns="http://www.w3.org/2000/svg"
                className="h-4 w-4"
                fill="none"
                viewBox="0 0 24 24"
                stroke="currentColor"
              >
                <title>Code icon</title>
                <path
                  strokeLinecap="round"
                  strokeLinejoin="round"
                  strokeWidth={2}
                  d="M10 20l4-16m4 4l4 4-4 4M6 16l-4-4 4-4"
                />
              </svg>
              <span>Code</span>
            </button>
          </div>
        ) : (
          <div className="h-10"></div>
        )}
        {onShare ? (
          !showWelcome && (
            <div className="flex items-center gap-2">
              {shareStatus && (
                <div className="animate-fade-in bg-accent-00-light dark:bg-accent-00-dark text-light-primary dark:text-dark-primary rounded-lg px-3 py-1 text-sm">
                  {shareStatus}
                </div>
              )}
              <div className="bg-light-decorative-00 dark:bg-dark-decorative-00 flex space-x-1 rounded-lg p-1 shadow-sm">
                <button
                  type="button"
                  onClick={onShare}
                  className="text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01 flex items-center space-x-1.5 rounded-md px-4 py-1.5 text-sm font-medium transition-colors"
                  aria-label="Share app"
                >
                  <svg
                    xmlns="http://www.w3.org/2000/svg"
                    className="h-4 w-4"
                    fill="none"
                    viewBox="0 0 24 24"
                    stroke="currentColor"
                  >
                    <title>Share icon</title>
                    <path
                      strokeLinecap="round"
                      strokeLinejoin="round"
                      strokeWidth={2}
                      d="M8.684 13.342C8.886 12.938 9 12.482 9 12c0-.482-.114-.938-.316-1.342m0 2.684a3 3 0 110-2.684m0 2.684l6.632 3.316m-6.632-6l6.632-3.316m0 0a3 3 0 105.367-2.684 3 3 0 00-5.367 2.684zm0 9.316a3 3 0 105.368 2.684 3 3 0 00-5.368-2.684z"
                    />
                  </svg>
                  <span>Share</span>
                </button>
              </div>
            </div>
          )
        ) : (
          <div className="h-10 w-10"></div>
        )}
      </div>

      {showWelcome ? (
        <div className="h-full" style={{ height: 'calc(100vh - 49px)' }}>
          <WelcomeScreen />
        </div>
      ) : (
        <div data-testid="sandpack-provider">
          <SandpackProvider
            key={sandpackKey}
            template="vite-react"
            options={{
              externalResources: ['https://cdn.tailwindcss.com'],
              classes: { 'sp-wrapper': 'h-full' },
            }}
            customSetup={{
              dependencies: {
                'use-fireproof': '0.20.0-dev-preview-52',
                ...(dependencies || {}),
              },
            }}
            files={filesRef.current}
            theme={isDarkMode ? 'dark' : 'light'}
          >
            <SandpackEventListener
              setActiveView={(view) => {
                if (!lockCodeView) {
                  setActiveView(view);
                }
              }}
              setBundlingComplete={setBundlingComplete}
              isStreaming={hasStreamingContent}
              onScreenshotCaptured={onScreenshotCaptured}
            />
            {hasStreamingContent && <SandpackScrollController isStreaming={hasStreamingContent} />}
            <SandpackLayout className="h-full" style={{ height: 'calc(100vh - 49px)' }}>
              <div
                style={{
                  display: activeView === 'preview' ? 'block' : 'none',
                  height: '100%',
                  width: '100%',
                }}
              >
                <SandpackPreview
                  showNavigator={false}
                  showOpenInCodeSandbox={false}
                  showRefreshButton={true}
                  showRestartButton={false}
                  showOpenNewtab={false}
                  className="h-full w-full"
                  style={{ height: '100%' }}
                />
              </div>
              <div
                style={{
                  display: activeView === 'code' ? 'block' : 'none',
                  height: '100%',
                  width: '100%',
                }}
                ref={codeEditorRef}
              >
                <SandpackCodeEditor
                  style={{ height: '100%' }}
                  showLineNumbers={false}
                  wrapContent
                  showInlineErrors
                />
              </div>
            </SandpackLayout>
          </SandpackProvider>
        </div>
      )}

      <div className="result-content">
        {!showWelcome && (
          <button
            data-testid="copy-button"
            onClick={() => navigator.clipboard.writeText(displayCode)}
            className="text-light-primary dark:text-dark-primary hover:bg-light-decorative-01 dark:hover:bg-dark-decorative-01 rounded-md px-4 py-1.5 text-sm font-medium transition-colors"
          >
            Copy to Clipboard
          </button>
        )}
        {streamingCode ? (
          <div>{currentStreamContent}</div>
        ) : (
          <div>{completedMessage || currentMessage?.content || ''}</div>
        )}
      </div>
    </div>
  );
}

function cleanCodeBeforeImport(codeString: string) {
  return codeString.replace(/^[\s\S]*?(import|export)/, '$1');
}

export default ResultPreview;
