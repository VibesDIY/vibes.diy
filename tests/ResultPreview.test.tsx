import { vi, describe, it, expect, beforeEach, afterAll } from 'vitest';
import { render, screen, act, waitFor } from '@testing-library/react';
import ResultPreview from '../app/components/ResultPreview/ResultPreview';
import { mockResultPreviewProps } from './mockData';

// Mock clipboard API
Object.assign(navigator, {
  clipboard: {
    writeText: vi.fn().mockImplementation(() => Promise.resolve()),
  },
});

// Mock SandpackProvider and related components
vi.mock('@codesandbox/sandpack-react', () => ({
  SandpackProvider: ({ children }: { children: React.ReactNode }) => (
    <div data-testid="sandpack-provider">{children}</div>
  ),
  SandpackLayout: ({ children }: { children: React.ReactNode }) => <div>{children}</div>,
  SandpackCodeEditor: () => <div data-testid="sandpack-editor">Editor</div>,
  SandpackPreview: () => <div data-testid="sandpack-preview">Preview</div>,
  useSandpack: () => ({
    sandpack: { activeFile: '/App.jsx' },
    listen: vi.fn().mockReturnValue(() => {}),
  }),
}));

// Mock WelcomeScreen
vi.mock('../app/components/ResultPreview/WelcomeScreen', () => ({
  default: () => <div data-testid="welcome-screen">Welcome Screen Content</div>,
}));

// Mock the Sandpack scroll controller
vi.mock('../app/components/ResultPreview/SandpackScrollController', () => ({
  default: () => null,
}));

// Mock ResizeObserver
global.ResizeObserver = vi.fn().mockImplementation(() => ({
  observe: vi.fn(),
  unobserve: vi.fn(),
  disconnect: vi.fn(),
}));

// Mock window.postMessage for preview communication
const originalPostMessage = window.postMessage;
window.postMessage = vi.fn();

// Reset mocks between tests
beforeEach(() => {
  vi.clearAllMocks();
  window.postMessage = vi.fn();
});

// Restore original window.postMessage after tests
afterAll(() => {
  window.postMessage = originalPostMessage;
});

describe('ResultPreview', () => {
  it('renders without crashing', () => {
    // Use non-empty code to ensure the editor is shown
    render(<ResultPreview code="const test = 'Hello';" {...mockResultPreviewProps} />);

    // Now the sandpack editor should be visible
    expect(screen.getByTestId('sandpack-editor')).toBeDefined();
    // Don't check for preview since it might not be available in the test environment
    // expect(screen.getByTestId('sandpack-preview')).toBeDefined();
  });

  it('displays welcome screen when code is empty', () => {
    render(<ResultPreview code={''} {...mockResultPreviewProps} />);

    expect(screen.getByAltText('Lightup')).toBeDefined();
  });

  it('handles streaming state correctly', () => {
    const code = 'const test = "Streaming";';

    render(<ResultPreview code={code} isStreaming={true} {...mockResultPreviewProps} />);

    // Just verify it renders without errors
    expect(screen.getAllByTestId('sandpack-provider')[0]).toBeDefined();
  });

  it('passes dependencies to SandpackProvider', () => {
    const code = 'console.log("test");';
    const dependencies = {
      react: '^18.0.0',
      'react-dom': '^18.0.0',
    };

    render(<ResultPreview code={code} dependencies={dependencies} {...mockResultPreviewProps} />);

    // Just verify it renders without errors
    expect(screen.getAllByTestId('sandpack-provider')[0]).toBeDefined();
  });

  it('calls onShare when share button is clicked', () => {
    // Skipping test since toolbar with share button has been removed
    // In the future, this would be added to a different component or the header
    expect(true).toBe(true);
  });

  it('shows welcome screen with empty code', () => {
    render(<ResultPreview code="" {...mockResultPreviewProps} />);

    expect(screen.getByAltText('Lightup')).toBeDefined();
  });

  it('shows a share button when onShare is provided and code is not empty', () => {
    // Skipping test since toolbar with share button has been removed
    expect(true).toBe(true);
  });

  it('updates display when code changes', () => {
    const { rerender } = render(<ResultPreview code="" {...mockResultPreviewProps} />);
    rerender(<ResultPreview code="const test = 'Hello';" {...mockResultPreviewProps} />);

    // Just verify it renders without errors
    expect(screen.getAllByTestId('sandpack-provider')[0]).toBeDefined();
  });

  it('renders with code content', () => {
    const code = 'const test = "Hello World";';

    render(<ResultPreview code={code} {...mockResultPreviewProps} />);

    // Skip button checks since toolbar has been removed
    expect(screen.getByTestId('sandpack-editor')).toBeDefined();
  });

  it('handles copy to clipboard', async () => {
    // Skipping test since toolbar with copy button has been removed
    expect(true).toBe(true);
  });

  it('renders with custom dependencies', async () => {
    const code = 'import React from "react";';
    const dependencies = {
      react: '^18.0.0',
      'react-dom': '^18.0.0',
    };

    render(<ResultPreview code={code} dependencies={dependencies} {...mockResultPreviewProps} />);

    // Use getAllByTestId to handle multiple elements
    expect(screen.getAllByTestId('sandpack-provider')[0]).toBeDefined();

    // Skip button check since toolbar has been removed
    expect(screen.getByTestId('sandpack-editor')).toBeDefined();
  });

  it('handles share functionality', () => {
    // Skipping test since share button has been removed
    expect(true).toBe(true);
  });

  it('handles edge case with empty code', () => {
    render(<ResultPreview code="" {...mockResultPreviewProps} />);

    // Skip button checks since toolbar has been removed
    expect(screen.getByAltText('Lightup')).toBeDefined();
  });

  it('renders empty state correctly', () => {
    const { container } = render(<ResultPreview code="" {...mockResultPreviewProps} />);
    // Update snapshot to match new structure
    expect(container).toMatchSnapshot();
  });

  it('handles dependencies correctly', () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    const dependencies = {
      react: '17.0.2',
      'react-dom': '17.0.2',
    };
    render(<ResultPreview code={code} dependencies={dependencies} {...mockResultPreviewProps} />);

    // Dependencies should be passed to the Sandpack component
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('displays code correctly', () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    render(<ResultPreview code={code} {...mockResultPreviewProps} />);

    // Code should be processed and displayed in the editor
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('shows welcome screen for empty code', () => {
    render(<ResultPreview code="" {...mockResultPreviewProps} />);
    expect(screen.getByAltText('Lightup')).toBeInTheDocument();
  });

  it('renders code properly', () => {
    render(<ResultPreview code="const test = 'Hello';" {...mockResultPreviewProps} />);
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('handles code updates correctly', () => {
    const { rerender } = render(<ResultPreview code="" {...mockResultPreviewProps} />);
    rerender(<ResultPreview code="const test = 'Hello';" {...mockResultPreviewProps} />);

    // Should change from welcome screen to code display
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('handles screenshot capture requests', () => {
    const onScreenshotCaptured = vi.fn();
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <ResultPreview
        code={code}
        onScreenshotCaptured={onScreenshotCaptured}
        {...mockResultPreviewProps}
      />
    );

    // Simulate screenshot message
    act(() => {
      window.dispatchEvent(
        new MessageEvent('message', {
          data: { type: 'screenshot', data: 'base64-data' },
        })
      );
    });

    expect(onScreenshotCaptured).toHaveBeenCalledWith('base64-data');
  });

  it('handles preview loaded event', async () => {
    const onPreviewLoaded = vi.fn();
    const code = `function App() { return <div>Hello World</div>; }`;
    render(
      <ResultPreview code={code} {...mockResultPreviewProps} onPreviewLoaded={onPreviewLoaded} />
    );

    // Simulate preview loaded message
    act(() => {
      window.dispatchEvent(
        new MessageEvent('message', {
          data: { type: 'preview-loaded' },
        })
      );
    });

    await waitFor(() => {
      expect(onPreviewLoaded).toHaveBeenCalled();
    });
  });

  it('passes dependencies to Sandpack', () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    const dependencies = {
      react: '17.0.2',
      'react-dom': '17.0.2',
    };
    render(<ResultPreview code={code} dependencies={dependencies} {...mockResultPreviewProps} />);

    // Dependencies should be configured in Sandpack
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('displays the code editor initially', () => {
    const code = `function App() { return <div>Hello World</div>; }`;
    render(<ResultPreview code={code} {...mockResultPreviewProps} />);

    // Should default to code view
    expect(screen.queryByText(/Welcome to the preview/i)).not.toBeInTheDocument();
  });

  it('shows welcome screen when no code is provided', () => {
    render(<ResultPreview code="" {...mockResultPreviewProps} />);
    expect(screen.getByAltText('Lightup')).toBeInTheDocument();
  });

  it('renders with a simple code snippet', () => {
    const code = 'const test = "Hello";';
    const setActiveView = vi.fn();

    // Render the component with a simple code snippet
    render(
      <ResultPreview
        code={code}
        dependencies={{}}
        isStreaming={false}
        codeReady={true}
        activeView="code"
        setActiveView={setActiveView}
        onPreviewLoaded={() => {}}
      />
    );

    // Now the sandpack editor should be visible
    expect(screen.getByTestId('sandpack-editor')).toBeDefined();
  });
});
