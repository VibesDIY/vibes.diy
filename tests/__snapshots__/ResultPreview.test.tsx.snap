// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ResultPreview > renders empty state correctly 1`] = `
<div>
  <div
    class="h-full"
    style="overflow: hidden;"
  >
    <style>
      
  @keyframes spin-slow {
    from {
      transform: rotate(0deg);
    }
    to {
      transform: rotate(360deg);
    }
  }
  .animate-spin-slow {
    animation: spin-slow 1s linear infinite;
  }

    </style>
    <div
      class="h-full"
    >
      <div
        class="bg-light-background-00 dark:bg-dark-background-00 flex h-full flex-col items-center justify-center"
      >
        <img
          alt="Lightup"
          class="logo-pulse h-auto w-full max-w-xs opacity-10"
          src="/lightup.png"
        />
      </div>
    </div>
  </div>
</div>
`;
